les/lespay/server/balance.go

```diff
@@ -543,8 +320,7 @@ func (n *NodeBalance) deactivate(close bool) {
 		n.updateEvent = nil
 	}
 	n.storeBalance(true, true)
-	n.bt.activeBalance.SubExp(n.balance.pos)
-	n.bt.inactiveBalance.AddExp(n.balance.pos)
+	n.bt.switchBalance(n.balance.pos, false)
```

Comment:
The locking order is incorrect here (my version wasn't right either, I forgot to lock at all). You are locking `bt` while `n` is locked. `BalanceTracker` uses locked `NodeBalance` functions so if they both need to be locked then you should lock `bt` first. This is problematic to do with this `switchBalance` function. Calling it after `n.lock.Unlock()` (like you do with `adjustBalance`) is also dangerous because a total token amount calculation might be called after `n` is unlocked and before `bt` is locked, which can then give an incorrect result. If you want to make this adjustment nice and safe and automatic, this is one idea I have:
```
func (bt *BalanceTracker) updateTotalBalance(n *NodeBalance, callback func()) {
	bt.lock.Lock()
	n.lock.Lock()
	if n.active {
		bt.active.SubExp(n.balance.pos)
	} else {
		bt.inactive.SubExp(n.balance.pos)
	}
	callback()
	if n.active {
		bt.active.AddExp(n.balance.pos)
	} else {
		bt.inactive.AddExp(n.balance.pos)
	}
	n.lock.Unlock()
	bt.lock.Unlock()
}
```
So you can do the balance and active status update in the callback and you can always be sure that it does what needs to be done.

Reply:
> You are locking bt while n is locked. BalanceTracker uses locked NodeBalance functions so if they both need to be locked then you should lock bt first

I don't understand why we need to lock bt first. What do you mean `BalanceTracker uses locked NodeBalance functions`? 

The original idea for refactoring here is I don't want to mix the tracker and balances together. They actually are two separate components and tracker is managing a batch of balance entries. 

So for example when we use `storeBalance` in the tracker, I think we should encapsulate a function that will lock the node balance in side instead of locking in the tracker side.

Reply:
If we are using two mutexes A and B which can also be locked together then we should always do the locking in the same order, otherwise a deadlock can happen. I am fine with changing the locking scheme if it makes the code nicer but I think now we have such a potential deadlock which needs to be fixed. If `n.deactivate()` and `bt.TotalTokenAmount()` are called at the same time, `deactivate` can block waiting on `bt` when `n.bt.switchBalance(n.balance.pos, false)` is called while `TotalTokenAmount` can block at `n.GetRawBalance()`.

Reply:
Right, fixed